src/hdmf/common/table.py

```diff
@@ -79,46 +80,91 @@ class VectorIndex(Index):
     def __init__(self, **kwargs):
         call_docval_func(super().__init__, kwargs)
         self.target = getargs('target', kwargs)
+        self.__uint = np.uint8
+        self.__maxval = 255
+        if isinstance(self.data, (list, np.ndarray)):
+            if len(self.data) > 0:
+                self.__check_precision(len(self.target))
+            # adjust precision for types that we can adjust precision for
+            self.__adjust_precision(self.__uint)
 
     def add_vector(self, arg):
         """
         Add the given data value to the target VectorData and append the corresponding index to this VectorIndex
         :param arg: The data value to be added to self.target
         """
         self.target.extend(arg)
-        self.append(len(self.target))
+        self.append(self.__check_precision(len(self.target)))
+
+    def __check_precision(self, idx):
+        """
+        Check precision of current dataset and, if
+        necessary, adjust precision to accomodate new value.
+
+        Returns:
+            unsigned integer encoding of idx
+        """
+        if idx > self.__maxval:
+            nbits = (np.log2(self.__maxval + 1)*2)
+            self.__uint = np.dtype('uint%d' % nbits).type
+            self.__maxval = 2**nbits - 1
+            self.__adjust_precision(self.__uint)
+        return self.__uint(idx)
+
+    def __adjust_precision(self, uint):
```

Comment:
@rly I added this because precision was not automatically getting adjusted. According to this function, it should have been.

https://github.com/hdmf-dev/hdmf/blob/712a50016fdcd015d9cd712b6dcc8434b6487c92/src/hdmf/build/objectmapper.py#L122-L129

This was the cause of failure of this test:
https://app.circleci.com/pipelines/github/hdmf-dev/hdmf/1489/workflows/d85924fb-11ac-455c-a2bd-31322ab95065/jobs/17977

Reply:
I think we would still want to store uints in the `VectorIndex` to avoid the conversion warning (or we could just remove the warning), so all of this uint code would still be used...